File: python/problems/leetcode/0088_merge_sorted_array/solution.py
"""LeetCode 88: Merge Sorted Array."""

from __future__ import annotations


class Solution:
    def merge(
        self, nums1: list[int], m: int, nums2: list[int], n: int
    ) -> None:
        i, j, k = m - 1, n - 1, m + n - 1
        while j >= 0:
            if i >= 0 and nums1[i] > nums2[j]:
                nums1[k] = nums1[i]
                i -= 1
            else:
                nums1[k] = nums2[j]
                j -= 1
            k -= 1


if __name__ == "__main__":
    a = [1, 2, 3, 0, 0, 0]
    Solution().merge(a, 3, [2, 5, 6], 3)
    assert a == [1, 2, 2, 3, 5, 6]
    b = [1]
    Solution().merge(b, 1, [], 0)
    assert b == [1]
    c = [0]
    Solution().merge(c, 0, [1], 1)
    assert c == [1]
    print("0088 OK")

File: cpp/problems/leetcode/0088_merge_sorted_array/solution.cpp
// LeetCode 88
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int i = m - 1, j = n - 1, k = m + n - 1;
        while (j >= 0) {
            if (i >= 0 && nums1[i] > nums2[j])
                nums1[k--] = nums1[i--];
            else
                nums1[k--] = nums2[j--];
        }
    }
};

int main() {
    vector<int> a{1, 2, 3, 0, 0, 0};
    vector<int> b{2, 5, 6};
    Solution().merge(a, 3, b, 3);
    assert((a == vector<int>{1, 2, 2, 3, 5, 6}));
    vector<int> c{1};
    vector<int> d{};
    Solution().merge(c, 1, d, 0);
    assert((c == vector<int>{1}));
    vector<int> e{0};
    vector<int> f{1};
    Solution().merge(e, 0, f, 1);
    assert((e == vector<int>{1}));
    cout << "0088 OK" << endl;
    return 0;
}
